### PR TITLE
Terraform: Add Application VMs to generated Ansible inventory

### LINKED ISSUE #558: Application Tier needs adding to the inventory
# Problem statement

The infrastructure generated by the Application tier Terraform code needs outputting in an Ansible inventory to enable further configuration.

### PR DESCRIPTION
**Pull request by peterchoo**
# Problem

See #558 - Application VMs are not currently in the generated Ansible inventory.  

## Description

This PR:

- Adds the Application VMs to the inventory
- Creates a duplicate YAML format inventory
- Updates the Ansible execution code to use the YAML format inventory
- Updates the Ansible Execution documentation to use the YAML format inventory
- Updates the HDB Loadbalancer use in `output_files` to prevent issue if multiple HANA platform entries are added to `databases`

Closes #558

## Notes

- [This change to the SAP Playbook](https://github.com/Azure/sap-hana/pull/560/files#diff-a73b84fbeab05fe149919c634ed1dc5aR9) is necessary as using a YAML format Inventory prevents [`inventory_dir` variable isn't always declared](https://github.com/ansible/ansible/issues/30901#issuecomment-334530854)

### CODE REVIEW

File: deploy/ansible/sap_playbook.yml [hunk 1]
```diff
@@ -6,7 +6,7 @@
   pre_tasks:
     - name: Include output JSON
       include_vars:
-        file: "{{ inventory_dir }}/output.json"
+        file: "{{ lookup('env', 'PWD') }}/output.json"
```
Comment (rukawata) [inline deploy/ansible/sap_playbook.yml:9]: I think the problem here is not the yml format, rather missing localhost definition in the yml.
I tested by adding below part in the hosts.yml and it is fine to use `inventory_dir`.
```yml
    local:
      hosts:
        localhost:
          ansible_connection: "local"
          ansible_user: "azureadm"
```

One disadvantage of using `{{ lookup('env', 'PWD') }}` is that when ansible-playbook is executed somewhere else other than `/home/<rti_user>`

File: deploy/terraform/modules/output_files/inventory.tf [hunk 2]
```diff
@@ -83,8 +83,31 @@ resource "local_file" "ansible-inventory" {
...
+    ips-jumpboxes-linux   = local.ips-jumpboxes-linux,
+    ips-dbnodes-admin     = local.ips-dbnodes-admin,
+    ips-dbnodes-db        = local.ips-dbnodes-db,
+    dbnodes               = local.dbnodes,
+    application           = var.application,
+    ips-scs               = local.ips-scs,
+    ips-app               = local.ips-app
+    }
+  )
+  filename = "${terraform.workspace}/ansible_config_files/hosts.yml"
```
Comment (rukawata) [inline deploy/terraform/modules/output_files/inventory.tf:112]: A trigger is missing in [ansible.tf](https://github.com/Azure/sap-hana/blob/1f3bd52b00bb415234eb7926915f87b1725e4793/deploy/terraform/ansible.tf#L9-L11).
This is to trigger the copy over to RTI when there is changes in the inventory (eg. switch from single node HANA to HA).

File: deploy/terraform/modules/output_files/ansible_inventory.yml.tmpl [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+      children:
+        jumpboxes_windows:
+        jumpboxes_linux:
+
+    all_linux_servers:
+      children:
+        jumpboxes_linux:
+        hanadbnodes:
+        scs:
+        app:
```
Comment (YunziZhangAzure) [inline deploy/terraform/modules/output_files/ansible_inventory.yml.tmpl:91]: For host.yml, should we include SID information?

Comment (rukawata) [reply deploy/terraform/modules/output_files/ansible_inventory.yml.tmpl:91]: To add SID will break the existing playbook unless we still use hosts in the ansible-play part for now. 
Option 1: have SID in the inventory yml, use inventory yml, change ansible code.
Option 2: have SID in the inventory yml, use inventory hosts, no change in ansible.

I will leave it up to @peterchoo to decide.

File: deploy/ansible/sap_playbook.yml (hunk 1 above)
Comment (YunziZhangAzure) [reply deploy/ansible/sap_playbook.yml:9]: Yes, I find the same thing.

Comment (peterchoo) [reply deploy/ansible/sap_playbook.yml:9]: I think this is the better solution. I'll update with this

File: deploy/terraform/modules/output_files/ansible_inventory.yml.tmpl [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    iscsi:
+      hosts:
+%{~ for ip-iscsi in ips-iscsi }
+        ${ip-iscsi}:
+%{~ if iscsi.authentication.type == "key" }
+          ansible_connection: "ssh"
+          ansible_user:       "${iscsi.authentication.username}"
+%{~ endif }
+%{~ if iscsi.authentication.type == "password" }
+          ansible_user:        "${iscsi.authentication.username}"
```
[the remark this answers is not in the document]
Comment (peterchoo) [reply deploy/terraform/modules/output_files/ansible_inventory.yml.tmpl:12]: Good spot, that had been my intention, will finish that bit off now! :)

File: deploy/terraform/modules/output_files/ansible_inventory.yml.tmpl [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    jumpboxes_windows:
+      hosts:
+%{~ for jumpbox-windows in jumpboxes-windows }
+        ${ips-jumpboxes-windows[index(jumpboxes-windows, jumpbox-windows)]}:
+          ansible_connection:                   "winrm"
+          ansible_user:                         "${jumpbox-windows.authentication.username}"
+          ansible_ssh_pass:                     "${jumpbox-windows.authentication.password}"
+          ansible_ssh_port:                     5986
+          ansible_winrm_transport:              "ntlm"
+          ansible_winrm_server_cert_validation: "ignore"
```
[the remark this answers is not in the document]
Comment (peterchoo) [reply deploy/terraform/modules/output_files/ansible_inventory.yml.tmpl:27]: This is a direct copy of the  options that are currently in the INI format inventory.

File: deploy/terraform/modules/output_files/inventory.tf (hunk 2 above)
Comment (peterchoo) [reply deploy/terraform/modules/output_files/inventory.tf:112]: I wasn't aware of that, I'll add it

File: deploy/terraform/modules/output_files/ansible_inventory.yml.tmpl (hunk 3 above)
Comment (peterchoo) [reply deploy/terraform/modules/output_files/ansible_inventory.yml.tmpl:91]: I think adding SID definition at this stage would be overkill for this particular PR.  As the current Ansible code is designed to work with only one HA pair, I think this change should not be included.

File: deploy/terraform/variables_local.tf [hunk 6]
```diff
@@ -23,5 +23,6 @@ locals {
   ], 0) : ""
 
   file_hosts  = fileexists("${terraform.workspace}/ansible_config_files/hosts") ? file("${terraform.workspace}/ansible_config_files/hosts") : ""
+  file_hosts2 = fileexists("${terraform.workspace}/ansible_config_files/hosts.yml") ? file("${terraform.workspace}/ansible_config_files/hosts.yml") : ""
```
Comment (rukawata) [inline deploy/terraform/variables_local.tf:26]: Can we rename it to `file_hosts_yml`?

File: deploy/terraform/ansible.tf [hunk 7]
```diff
@@ -8,6 +8,7 @@ resource "null_resource" "prepare_rti_files" {
 
   triggers = {
     hosts  = sha1(local.file_hosts)
+    hosts2 = sha1(local.file_hosts2)
```
Comment (rukawata) [inline deploy/terraform/ansible.tf:11]: Same as above:
```suggestion
    hosts_yml = sha1(local.file_hosts_yml)
```

File: deploy/terraform/modules/output_files/ansible_inventory.tmpl [hunk 8]
```diff
@@ -71,3 +77,13 @@ ${ip-dbnode-admin}  ansible_connection=ssh  ansible_user=${dbnodes[index(ips-dbn
...
 %{~ endif }
 %{~ endfor }
+
+[scs]
+%{~ for ip-scs in ips-scs }
+${ip-scs}  ansible_connection=ssh ansible_user=${application.authentication.username}
+%{~ endfor }
+
+[app]
+%{ for ip-app in ips-app }
```
Comment (YunziZhangAzure) [inline deploy/terraform/modules/output_files/ansible_inventory.tmpl:87]: %{ for ip-app in ips-app }
```suggestion
%{~ for ip-app in ips-app }
```

**Pull request merged**